File: 2026/07-July/08-Wednesday/solution.py
class Solution:
    def concatNonZero(self, s: str, queries: List[List[int]]) -> List[int]:
        MOD = 1_000_000_007
        n = len(s)

        digits = []
        positions = []

        for i, ch in enumerate(s):
            if ch != '0':
                digits.append(ord(ch) - ord('0'))
                positions.append(i)

        m = len(digits)

        if m == 0:
            return [0] * len(queries)

        power10 = [1] * (m + 1)
        for i in range(1, m + 1):
            power10[i] = (power10[i - 1] * 10) % MOD

        prefix_num = [0] * (m + 1)
        prefix_sum = [0] * (m + 1)

        for i in range(m):
            prefix_num[i + 1] = (prefix_num[i] * 10 + digits[i]) % MOD
            prefix_sum[i + 1] = prefix_sum[i] + digits[i]

        first_non_zero = [m] * (n + 1)
        idx = m - 1
        for i in range(n - 1, -1, -1):
            if idx >= 0 and positions[idx] == i:
                first_non_zero[i] = idx
                idx -= 1
            else:
                first_non_zero[i] = first_non_zero[i + 1]

        last_non_zero = [-1] * n
        idx = 0
        current = -1
        for i in range(n):
            if idx < m and positions[idx] == i:
                current = idx
                idx += 1
            last_non_zero[i] = current

        answer = []

        for left, right in queries:
            l = first_non_zero[left]
            r = last_non_zero[right]

            if l > r or l == m or r == -1:
                answer.append(0)
                continue

            length = r - l + 1

            value = (
                prefix_num[r + 1]
                - prefix_num[l] * power10[length]
            ) % MOD

            digit_sum = prefix_sum[r + 1] - prefix_sum[l]

            answer.append((value * digit_sum) % MOD)

        return answer

File: 2026/07-July/08-Wednesday/solution.cpp
class Solution {
public:
    vector<int> sumAndMultiply(string s, vector<vector<int>>& queries) {
        const long long MOD = 1000000007LL;

        int n = s.size();

        vector<int> digits;
        vector<int> positions;

        for (int i = 0; i < n; i++) {
            if (s[i] != '0') {
                digits.push_back(s[i] - '0');
                positions.push_back(i);
            }
        }

        int m = digits.size();
        vector<int> ans(queries.size(), 0);

        if (m == 0)
            return ans;

        vector<long long> power10(m + 1, 1);
        for (int i = 1; i <= m; i++)
            power10[i] = (power10[i - 1] * 10) % MOD;

        vector<long long> prefixValue(m + 1, 0);
        vector<long long> prefixSum(m + 1, 0);

        for (int i = 0; i < m; i++) {
            prefixValue[i + 1] = (prefixValue[i] * 10 + digits[i]) % MOD;
            prefixSum[i + 1] = prefixSum[i] + digits[i];
        }

        vector<int> first(n + 1, m);

        int idx = m - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (idx >= 0 && positions[idx] == i) {
                first[i] = idx;
                idx--;
            } else {
                first[i] = first[i + 1];
            }
        }

        vector<int> last(n, -1);

        idx = 0;
        int current = -1;

        for (int i = 0; i < n; i++) {
            if (idx < m && positions[idx] == i) {
                current = idx;
                idx++;
            }
            last[i] = current;
        }

        for (int i = 0; i < (int)queries.size(); i++) {
            int left = first[queries[i][0]];
            int right = last[queries[i][1]];

            if (left > right || left == m || right == -1) {
                ans[i] = 0;
                continue;
            }

            int len = right - left + 1;

            long long number =
                (prefixValue[right + 1] -
                 (prefixValue[left] * power10[len]) % MOD +
                 MOD) % MOD;

            long long digitSum = prefixSum[right + 1] - prefixSum[left];

            ans[i] = (number * digitSum) % MOD;
        }

        return ans;
    }
};
